### ISSUE TITLE
NGramsForecastAndSample and NGramsForecast unusable with Vietnamese 

### ISSUE THREAD

**Issue by lamhoangtung** [githubharald/CTCWordBeamSearch#8]
First, thanks for the great works, Harald. I’m trying to implement your Word Beam Search Algorithm to Vietnamese language. I modified the `charList`, `wordcharList` and the `corpus` for my need. It's been working good for `Words` and `NGrams` mode. But when I tried to use it in either `NGramsForecast` or `NGramsForecastAndSample`, it raised this error.

```
Python(23489,0x70000ee70000) malloc: *** error for object 0x7fa625811e00: incorrect checksum for freed object — object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
Python(23489,0x70000ef76000) malloc: *** error for object 0x7fa62714f940: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
zsh: abort python3 main.py — validate — wordbeamsearch
```

Which i have no ideal how to resolve.  I'm using `Tensorflow 1.7.0` on `MacOS 10.13.5`, `Ubuntu 16.04` and `18.04`

BTW, you can get a sense of Vietnamese language, its structure via this wiki page: https://en.wikipedia.org/wiki/Vietnamese_language

So I really need your help here. What did i’ve missed ? What thing i can do to resolve this problem ?

Thanks you.

**Comment by lamhoangtung** [comment]
It's worked when using the original files you provided. Here is all the files that I have changed:
- [corpus.txt](https://github.com/githubharald/CTCWordBeamSearch/files/2433787/corpus.txt)
- [charList.txt](https://github.com/githubharald/CTCWordBeamSearch/files/2433784/charList.txt)
- [wordCharList.txt](https://github.com/githubharald/CTCWordBeamSearch/files/2433785/wordCharList.txt)

**Comment by githubharald** [comment]
and I also need at least one pair of mat_X.csv and gt_X.txt files please (where X is some number like 0, 1, ...).

**Comment by lamhoangtung** [comment]
How can i create the .csv files for you ? Just dump the output tensor from RNN for you?

Anyways, here is the test image and ground of truth text

- [gt_0.txt](https://github.com/githubharald/CTCWordBeamSearch/files/2437575/gt_0.txt): "770/2 Bình Giã, Phường 10, Thành Phố Vũng Tàu, Bà Rịa - Vũng Tàu"
- ![img_0](https://user-images.githubusercontent.com/26834698/46345385-9303c480-c66e-11e8-9763-18bf8e5ad79f.png)

**Comment by githubharald** [comment]
yes, just dump the RNN output for the sample you have shown in your post.
Should be of 2d shape and size (C+1) x T, where C is number of chars + blank char and T is sequence length.
With or without softmax applied, that doesn't matter (just tell me if it is applied or not).

**Comment by githubharald** [comment]
do you use the SimpleHTR model for your text recognition task? If yes I can provide the code to dump the RNN output of a certain batch element.

**Comment by githubharald** [comment]
1. in method `__init__` add this line such that the output of the RNN can be accessed from other methods too by putting it into self:
```python
self.rnnOutput = tf.transpose(rnnOut3d, [1, 0, 2])
```



2. in method `inferBatch` put this code which additionally computes the RNN output and then dumps it into a CSV file:
```python
def inferBatch(self, batch):
	"feed a batch into the NN to recngnize the texts"
	decoded, rnnOutput = self.sess.run([self.decoder, self.rnnOutput], { self.inputImgs : batch.imgs, self.seqLen : [Model.maxTextLen] * Model.batchSize } )
	
	s=rnnOutput.shape
	b = 0 # first batch element
	csv='' # csv string
	for t in range(s[0]):
		for c in range(s[2]):
			csv+=str(rnnOutput[t,b,c])+';'
		csv+='\n'
	open('mat_0.csv', 'w').write(csv)
	
	return self.decoderOutputToText(decoded)
```
	
If you now call python main.py to infer the text on the test.png, you will get a file mat_0.csv in your src dir.

**Comment by lamhoangtung** [comment]
Here you go:

- [gt_0.txt](https://github.com/githubharald/CTCWordBeamSearch/files/2441790/gt_0.txt)
- ![img_0](https://user-images.githubusercontent.com/26834698/46407516-6e285380-c739-11e8-8a2e-03879ecd1248.png)
- [mat_0.csv](https://drive.google.com/open?id=1lzKNj8r26AxVu9cXq0_UqCvxqo-7rz3u)

Github don't support .csv file via comment so I have to upload it to Google Drive.

BTW, have you ever get any strange behavior when running infer mode? Sometimes I get the same output for all input when inference, but when I tried to use validation mode or when I tried to feed different image to the infer batch then it will be fine (fill all the batch with the same img or with white or black image are all result in same output for any input).

**Comment by githubharald** [comment]
I get the following results with word beam search:
* Words mode: Xã 0/ Bình Gi, Phường 10, Thành Phố Vũng Tần, Bà Rịa, Vũng Trung
* Ngrams+Forecast+Sample mode: Xã0/ Bình Gi, Phường 10, Thành Phố Vũng Tần, Bà Rịa, Vũng Trung

Maybe there is something wrong when passing the input texts to the algorithm through your python code. Take care that strings are passed as UTF8 encoded, as the C++ code expects UTF8 strings.

P.S.: you should use best path decoding while training (because faster & results based solely on neural network output). Word beam search is for inferring.

**Comment by lamhoangtung** [comment]
> Maybe there is something wrong when passing the input texts to the algorithm through your python code. Take care that strings are passed as UTF8 encoded, as the C++ code expects UTF8 strings.

What do you mean by `the input texts to the algorithm`, are you pointing at the `charList` or `corpus`, ..etc ?

**Comment by githubharald** [comment]
look at the test code:

* Passing to C++ function: https://github.com/githubharald/CTCWordBeamSearch/blob/master/tf/testCustomOp.py#L23
* Reading files: https://github.com/githubharald/CTCWordBeamSearch/blob/master/tf/testCustomOp.py#L72

Take care that in your code everything passed to the C++ function is UTF8 encoded.

**Comment by lamhoangtung** [comment]
I make sure everything is UTF8 encoded but the problem still remain

**Comment by githubharald** [comment]
What about testCustomOp.py when using your data? Did it work? 
The output is shown below:
```
...
Real example:
Label string:  [103 115  13  11  12  41 150 158 147  12  53 148   7  12  84 147 188 173
 158 146  12  14  13   7  12  88 147 113 158 147  12  84 147 166  12 101
 186 158 146  12  88 125 158   7  12  41 113  12  86 153 111   7  12 101
 186 158 146  12  88 179 182 158 146 204 204 204 204 204 204 204 204 204
 204 204 204 204 204 204 204 204 204 204 204 204 204 204 204 204 204 204
 204 204 204 204 204 204 204 204 204 204]
Char string: "Xã0/ Bình Gi, Phường 10, Thành Phố Vũng Tần, Bà Rịa, Vũng Trung"
```

* Do you get the same output?
* Does it crash?

**Comment by githubharald** [comment]
then the question is: what is the difference between the code of testCustomOp.py and your neural network? Something must be different.

* Is the word_beam_search(...) function called with exactly the same arguments? Compare the arguments on byte-level.
* What is the shape of the input matrix in both cases?
* Is there only a single call of the function word_beam_search(...), as it is doing all the init-stuff and should only be called once.

**Comment by lamhoangtung** [comment]
After a lot of debugging and logging to both your C++ code and my python code, I found that the problem will disappear when I compile with single-thread decoding ...

**Comment by githubharald** [comment]
ok, thanks for your analysis.

**Comment by githubharald** [comment]
The fix for issue https://github.com/githubharald/CTCWordBeamSearch/issues/29 might also solve the problem you reported.
